### PR TITLE
Create a new Build for scratch builds

### PR DESCRIPTION
**Pull request by vrutkovs**
This creates a new build on existing BuildConfig (if found) instead of updating BuildConfig

TODO:
 * [x] Add a new config option for scratch builds (after build_types were removed)
 * [x] Add tests
 * [x] Verify on several existing projects

### CODE REVIEW

**Comment by twaugh** [comment]
Why look for an existing BuildConfig in that situation? We already have everything we need to submit a new build don't we?

**Comment by twaugh** [comment]
I think it might be better if `create_prod_build()` only called `self.os.create_build()` when scratch is set, and `_create_build_config_and_build()` otherwise.

**Comment by vrutkovs** [comment]
> We already have everything we need to submit a new build don't we?

Indeed, if we specify BuildConfig there the build will be displayed as belonging to this buildconfig. Otherwise it will create a new buildconfig

OTOH I'm not entirely sure about this - wouldn't it be better to create a new ```scratchbuilds``` buildconfig in case we'll decide to use the same namespace for scratch builds?

File: tests/test_api.py [hunk 1]
```diff
@@ -722,3 +729,41 @@ def test_create_build_config_auto_start(self):
...
+        updated_build_json['spec'] = {}
+        updated_build_json['spec']['serviceAccount'] = 'builder'
+        build_name = 'scratch-%s' % datetime.datetime.now().strftime('%Y%m%d%H%M%S')
+        updated_build_json['metadata']['name'] = build_name
+
+        (flexmock(osbs.os)
+            .should_receive('create_build')
+            .with_args(updated_build_json)
+            .once()
+            .and_return(flexmock(json=lambda: {'spam': 'maps'})))
```
Comment (twaugh) [inline tests/test_api.py]: Could add a flexmock(osbs.os).should_receive('create_build_config').never() to make that explicit.

Comment (vrutkovs) [inline tests/test_api.py]: Good catch, added

File: osbs/api.py [hunk 2]
```diff
@@ -381,10 +394,13 @@ def create_prod_build(self, git_uri, git_ref,
             git_push_url=self.build_conf.get_git_push_url(),
             git_push_username=self.build_conf.get_git_push_username(),
             builder_build_json_dir=self.build_conf.get_builder_build_json_store(),
-            scratch=scratch,
+            scratch=self.build_conf.get_scratch(),
```
Comment (twaugh) [inline osbs/api.py]: Careful here. This changes means that create_prod_build(..., scratch=True) will no longer have any effect. Instead, the scratch parameter would have to be passed to the Configuration() constructor, or put in a config file.

We could actually solve this problem by having get_scratch() take a parameter to use as the default, and giving it the 'scratch' we were passed here as its parameter.

File: osbs/api.py [hunk 3]
```diff
@@ -259,6 +260,18 @@ def _verify_no_running_builds(self, build_config_name):
                 msg = self._panic_msg_for_more_running_builds(build_config_name, running_builds)
             raise OsbsException(msg)
 
+    def _create_scratch_build(self, build_request):
+        build_json = build_request.render()
+        build_json['kind'] = 'Build'
+        if 'spec' not in build_json.keys():
+            build_json['spec'] = {}
+        build_json['spec']['serviceAccount'] = 'builder'
+        build_json['metadata']['labels']['scratch'] = 'true'
```
Comment (twaugh) [inline osbs/api.py]: +1

File: osbs/build/build_request.py [hunk 4]
```diff
@@ -50,7 +50,7 @@ def __init__(self, build_json_store):
         self._resource_limits = None
         self._openshift_required_version = parse_version('1.0.6')
         # For the koji "scratch" build type
-        self.scratch = False
+        self.scratch = None
```
Comment (twaugh) [inline osbs/build/build_request.py]: Curious -- why the change here?

Comment (twaugh) [inline osbs/build/build_request.py]: Oh, I get it, I think. Don't provide a value, so as to avoid overriding it if it is set elsewhere.

Comment (vrutkovs) [inline osbs/build/build_request.py]: This is done to make sure config param doesn't get overwritten by default params.

File: tests/test_api.py [hunk 5]
```diff
@@ -722,3 +729,49 @@ def test_create_build_config_auto_start(self):
 
         build_response = osbs._create_build_config_and_build(build_request)
         assert build_response.json == {'spam': 'maps'}
+
+    def test_scratch_build_config(self):
```
Comment (twaugh) [inline tests/test_api.py]: Could we add a test for passing 'scratch=True' to the create_build() API method?

Comment (vrutkovs) [inline tests/test_api.py]: Added ```test_scratch_param_to_create_build``` test case

**Comment by twaugh** [comment]
LGTM

File: osbs/cli/main.py [hunk 6]
```diff
@@ -566,7 +566,7 @@ def cli():
     build_parser.add_argument("--storage-limit", action='store', required=False,
                               help="storage limit")
     build_parser.add_argument("--scratch", action='store_true', required=False,
-                              default=False, help="perform a scratch build")
```
Comment (lcarva) [inline osbs/cli/main.py]: You could remove the `default` param as its default value is `None`.

**Comment by lcarva** [comment]
LGTM
